File: r-shiny/train.R
library(e1071)

#' Read the bananas data
x <- read.csv("r-shiny/bananas.csv")
x$ripeness <- factor(x$ripeness, c("Under", "Ripe", "Very", "Over"))

#' Multinomial classification with Support Vector Machines
m <- svm(ripeness ~ green + yellow + brown,
  data = x,
  probability = TRUE
)

#' Two-way table to test prediction accuracy
table(x$ripeness, predict(m))
sum(diag(table(x$ripeness, predict(m)))) / nrow(x)

#' Predict ripeness class
predict(m, data.frame(green = 1, yellow = 0, brown = 0), probability = TRUE)
predict(m, data.frame(green = 0, yellow = 1, brown = 0), probability = TRUE)
predict(m, data.frame(green = 0, yellow = 0, brown = 1), probability = TRUE)
predict(m, data.frame(green = 0.1, yellow = 0.2, brown = 0.7), probability = TRUE)

#' Save the model object
saveRDS(m, "r-shiny/bananas-svm.rds")

File: py-shiny/train.py
import pandas as pd
from joblib import dump
from sklearn import svm

# Global
x = pd.read_csv('py-shiny/bananas.csv')

# Train SVM
x.loc[x.ripeness == 'Under', 'target'] = 0
x.loc[x.ripeness == 'Ripe', 'target'] = 1
x.loc[x.ripeness == 'Very', 'target'] = 2
x.loc[x.ripeness == 'Over', 'target'] = 3
data_X = x[['green', 'yellow', 'brown']].to_numpy()
data_y = x.target.values
svm_model = svm.SVC(probability = True)
svm_model.fit(data_X, data_y)

#' Predict ripeness class
svm_model.predict_proba([[1, 0, 0]])
svm_model.predict_proba([[0, 1, 0]])
svm_model.predict_proba([[0, 0, 1]])
svm_model.predict_proba([[0.1, 0.2, 0.7]])

# Write model object to file
dump(svm_model, 'py-shiny/bananas-svm.joblib')
